### PR TITLE
Handle SNS notifications for emails sent

### PR DESCRIPTION
**Pull request by albertisfu**
This PR contains:
- Signals to handle:
 -Complaints notifications
 -Bounces notifications
 -Delivery notifications

- Tests to verify notification events are received and handled properly according to each type of event.

- Function to handle soft bounces notifications
-`EmailFlag` and `BackoffEvent` models were added:
  During tests, I realized that storing user_id in these models are not necessary because we are going to flag and ban emails, not users, and there could exist some conflict cases, for example when a user changes its email and then we receive a bounce or complaint related with the previous user email address we won't be able to assign this event to a user because the user's email address changed. 
  I think it was better to create a date_created field directly in the model instead of using  `AbstractDateTimeModel` because we are not going to use a `date_modified` field and also I think we don't need  index for these datetime fields.

- Tests to verify soft bounces function works properly under each type of soft bounce notifications and different scenarios.

### CODE REVIEW

**Comment by albertisfu** [comment]
@mlissner added a commit in this PR for phase 2: handle hard bounces and delivery notifications because it also contains some improvements for previous commits.

I noticed I didn't explain how exponential backoff is working:

- The general idea is to have an average max retry period of 5 days the exact period will depend on how fast SES sends bounce notifications.

- The first time we get a soft bounce notification we are going to wait for 2 hours before trying to send another email to the same email address.

- If after sending a new email we get another soft bounce notification for the same email address the next waiting period will be 4 hours, if continue bouncing the next waiting period will be 8, 16, 32, and 64 hours in the 5º retry, for a total of 126 hours (in average 5.25 days) since the initial bounce, after the 5º retry if we receive another bounce notification the email address is banned.

Noticed there are some conflicts to merge with main, I'll check them and solve them.

File: cl/users/models.py
Comment (mlissner) [inline cl/users/models.py:294]: You can just remove the `else` statement and outdent the `get_or_create` and this will be more clear. Also, please use `flag_type=EmailFlag.BAN` instead of spelling out the choice.

File: cl/users/models.py
Comment (mlissner) [inline cl/users/models.py:265]:
```suggestion
    """Ban any email address that receives a hard bounce.
    
    If the bounce is of an unexpected type, also log a warning.
    
```

File: cl/users/models.py
Comment (mlissner) [inline cl/users/models.py:267]: I think I'd convert this to snake_case instead of camelCase even though AWS prefers to camelCase.

File: cl/users/models.py
Comment (mlissner) [inline cl/users/models.py:300]: Should have a newline here:
```suggestion
    """Handle a soft bounce notification received from SNS
    
```

File: cl/users/models.py
Comment (mlissner) [inline cl/users/models.py:302]: Again, snake_case, not camelCase, please.

File: cl/users/models.py
Comment (mlissner) [inline cl/users/models.py:309]: Adding a comment referencing this issue or the other place we discussed this would be helpful here to help explain the different types and what we do. A long comment here might not hurt.

https://github.com/freelawproject/courtlistener/issues/1942

File: cl/users/models.py
Comment (mlissner) [inline cl/users/models.py:336]: I think flipping this is a bit more clear, so it reads as:

> If the time now is after the retry date

Instead of:

> If the retry date is before now

Same meaning, but the first one makes more sense to me. So:

```suggestion
                if now() >= next_retry_date:
```

It *might* be even more clear to do:

```
if now() < next_retry_date:
    # It's not yet time to retry. 
    continue
```

Then you can outdent all that follows. 

I'm also curious, are we missing the `else` statement here? What if it's not yet time to send the email? Do we save it or something?

File: cl/users/models.py
Comment (mlissner) [inline cl/users/models.py:362]: There's a race condition here, but I suppose it's fine. Easier to live with it than to fix it, probably.

File: cl/users/tests.py
Comment (mlissner) [inline cl/users/tests.py:341]: I'm constantly encouraging people to use Pathlib for things like this, but dont' feel obligated to upgrade this code. One day I'll get mean about it, I guess. :)

**Review by mlissner — CHANGES_REQUESTED** [review]
OK! Lots of little stuff, but nothing too big. In general, it looks great and I'm excited for the next stage to land. 

A few high-level items though:

1. If you can add admin files for the email models, that'd be great. 

2. Let's move all the email logic out of the models.py file. Maybe put it in cl.users.email_handlers.py?

3. We need a rebase, unfortunately, but hopefully that won't be too bad.

Generally looks good! Thank you, Alberto!

File: cl/users/urls.py
Comment (albertisfu) [reply cl/users/urls.py]: I agree, solved!

File: docker/django/Dockerfile
Comment (albertisfu) [reply docker/django/Dockerfile]: Done, I put it like this: 
`# For installing M2Crypto necessary for django-ses`
` libssl-dev swig`

File: cl/users/models.py
Comment (albertisfu) [reply cl/users/models.py]: Done! I added a class `SUB_TYPES` to handle these choices, also changed the field name from `reason` to `event_sub_type` because we are going to use for bounces and complaints subtypes

File: cl/users/models.py
Comment (albertisfu) [reply cl/users/models.py]: Thank you for the detailed explanation of why we need to use `AbstractDateTimeModel`.
I have updated the model to inherit from `AbstractDateTimeModel` and remove the old `date_created` field.

File: cl/users/models.py
Comment (albertisfu) [reply cl/users/models.py]: Done, I added a detailed comment.

File: cl/users/models.py
Comment (albertisfu) [reply cl/users/models.py:234]: Done, I added a detailed comment, like this:
```
 :email_address: an EmailFlag object belongs to an email address instead
    of a user, in this way, if users change their email address this won't
    affect new user email address.
```

File: cl/users/models.py
Comment (albertisfu) [reply cl/users/models.py]: Fixed, I use a SmallIntegerField for these choice fields.

File: cl/users/models.py
Comment (albertisfu) [reply cl/users/models.py]: Done, I added a comment like this: `The actual flag assigned, like: small_email_only` and also changed the field to be also an integer choice field, for now, we just use two flags "small_email_only" and "max_retry_reached"

File: cl/users/models.py
Comment (albertisfu) [reply cl/users/models.py]: Done, I added the comment.

File: cl/users/models.py
Comment (albertisfu) [reply cl/users/models.py]: Fixed! btw I moved all these email logic code to cl.users.email_handlers and cl.users.signals.py

File: cl/users/models.py
Comment (albertisfu) [reply cl/users/models.py]: Done!

File: cl/users/models.py
Comment (albertisfu) [reply cl/users/models.py]: Done!

File: cl/users/models.py
Comment (albertisfu) [reply cl/users/models.py]: Done! I added a detailed comment explaining how this works.

File: cl/users/models.py
Comment (albertisfu) [reply cl/users/models.py]: oh yes it's more clear like you suggested `if now() >= next_retry_date:` 

I used this option instead of `if now() < next_retry_date:` because here we need to act when the waiting period is over,  so we don't have an `else`, here is a screenshot where I show this part of the code, the `else` is "Don't update the backoff event" so we are just ignoring it, as you could see we have a `Queue Email` function before entering to these `if `statements, so yes we are going to queue this email to send it later.
 I put a comment in the code where we are going to call this function `#TODO Queue email function`

screen_1

File: cl/users/models.py
Comment (albertisfu) [reply cl/users/models.py]: I see, I definitely would like to fix this, would it be helpful to use `select_for_update()` and `@transaction.atomic`?
I would be very grateful if you could give me a hint as to whether this is a good approach or if there might be a better one.

File: cl/users/tests.py
Comment (albertisfu) [reply cl/users/tests.py]: Solved! I'm now using Pathlib :)

**Comment by albertisfu** [comment]
Thank you so much for your feedback Mike, I really appreciate it!

I think most of the changes are now solved, I left some comments explaining some of them, it is pending the one related to the race condition, also left a question about it.
Also solved:
- Moved email logic and signals outside of the models to cl.users.email_handlers and cl.users.signals.py
- Rebased with main, (poetry.lock) conflicts solved

File: cl/users/models.py
Comment (mlissner) [inline cl/users/models.py:218]: For the field level details, any reason to put it here instead of in the help_text or in a comment near the field? I think it moves the comment further from where you need it. It's also not how we usually have done it, but if this is a pattern elsewhere, I'd be happy to consider a different approach.

File: cl/users/models.py
Comment (mlissner) [inline cl/users/models.py:266]: I'd suggest moving this too, unless this is a common style I just haven't seen.

**Comment by mlissner** [comment]
Great, thank you. Just a couple more comments at the moment. Once you've got those in I'll do one more full review and hopefully we'll be good. I also wanted to share this with you:

https://github.com/django-ses/django-ses/issues/182

I think it'd be pretty easy to do, but I'm waiting for the maintainer to give us the OK. Once done, we can remove the docker tweaks, but I don't know when they'll reply, so I think we have to move forward with m2crypto for now.

**Comment by albertisfu** [comment]
I've solved the last two comments.

Wonderful I saw that you have already started working with `django-ses` contribution, if you need me to check something about it, let me know.

File: cl/users/models.py
Comment (mlissner) [reply cl/users/models.py]: Yes, I think `select_for_update` would be a simple and good way to do this.

File: cl/users/models.py
Comment (mlissner) [inline cl/users/models.py:53]: This should be closer to the model it's used in, not at the top of the file.

**Comment by mlissner** [comment]
I just caught one more little thing and I replied to the race condition thing. Otherwise, I think we're good without further review. Thank you!

**Comment by albertisfu** [comment]
- I've solved the race condition thing using `selec_for_update()` and `transaction.atomic`.

- I also added some constraints to models to make them unique and avoid duplicates, for `BackoffEvent` the constraint is by email_address.

- For `EmailFlag` the constraint is by email and `object_type = BAN`, this way it'll only be possible to have one BAN object per email_address, added a test to ensure we can update a banned email instead of creating a new object.

- I've rebased with main and solved poetry conflicts because I'm going also to rebase this with the custom backend email branch that I'm working on to test them together.

Seems there's an issue with pre-commit.ci check, but I think it came from main

**Comment by mlissner** [comment]
This looks ready to merge to me, but I think if we wait another 24 hours the new version of django-ses will remove the m2crypto junk, so I think we should wait for that, then fix the m2crypto import and then press ahead. Bill is also working on the docker image, so he might appreciate it not changing under him either, if we can help it. But no further tweaks on this from me.

**Comment by albertisfu** [comment]
I've added support for multiple addresses for `to`, `bcc`, `cc` and `reply_to` fields.

Now we can store and handle banned, small emails and backoff events for multiple recipients.

- When storing messages we can store the original `from` email address in this format: 
   `Admin User <success@simulator.amazonses.com>` or just as a simple email `address success@simulator.amazonses.com`

- For `to`, `bcc`, `cc` and `reply to` fields, we normalize the email addresses before sending and storing the message, an email address like this `["Admin User <success@simulator.amazonses.com>"]` turns to `  ["success@simulator.amazonses.com"]`

When sending to multiple recipients we apply the following actions:

- For banned email: we verify if a recipient is banned, if so we remove the banned email addresses from the list of recipients to avoid sending emails to those addresses, if all of the recipients are banned, we discard the message, we don't send or store it.

- For small email only: if at least one of the recipients requires a small version, we send the small email version to all recipients.

- For email addresses under a backoff waiting period: if we detect an email address that is under a backoff waiting period we remove that address from the recipient list to avoid sending to it, we'll queue the message for those recipients.

**Review by mlissner — CHANGES_REQUESTED** [review]
OK, this is ready except for *one* last thing. Can you please delete all your migrations and re-do them as a single one? Should be as simple as (roughly):

```
manage.py migrate users 0003
rm users/migrations/000{4..8}.*
manage.py migrate users -n add_email_handling
manage.py sqlmigrate users 0004
```

Migrations take time to run, so I try to avoid having more than one in a PR. They pile up!

**Comment by mlissner** [comment]
Thank you @albertisfu! Huge PR, and a lot of work. Our email system is going to be incredible.

**Pull request merged**
